### PR TITLE
New issuer

### PR DESCRIPTION
**Pull request by mattcollier**
Many points of discussion here.

### CODE REVIEW

**Comment by mattcollier** [comment]
Are we supposed to be saving issued credentials in an EDV?

**Comment by mattcollier** [comment]
@dlongley @gannan08 ready for for review.  Still a couple fixme's that need to be resolved.

Also, credentials issued are not visible on the wallet.  I'm not sure if that's because of an index mismatch, an issue with the `recipients` on the document or just that the credential displayer doesn't know what to do with the sample credential.  I'm going to be looking into this further.

**Comment by gannan08** [comment]
> @dlongley @gannan08 ready for for review. Still a couple fixme's that need to be resolved.
> 
> Also, credentials issued are not visible on the wallet. I'm not sure if that's because of an index mismatch, an issue with the `recipients` on the document or just that the credential displayer doesn't know what to do with the sample credential. I'm going to be looking into this further.

@mattcollier  The credentials that are issued are stored in the Issuer Instance's Credential EDV at the Wallet's EDV Service. The credentials that the Wallet displays are pulled from the Profile's Credential EDV at that Wallet's EDV Service.

In regards to issued credentials, I am unsure how we want to surface these credentials in the wallet -- if that's even something we want to do.

File: lib/http.js [hunk 1]
```diff
@@ -272,38 +273,29 @@ bedrock.events.on('bedrock-express.configure.routes', app => {
...
-
-      // get instance for issuer
-      const {instance} = await vcIssuer.instances.get(issuer);
-      const {controller: accountId} = instance;
+        body: {credential},
+        headers: {authorization},
+      } = req;
+
+      if(!(authorization && typeof authorization === 'string' &&
+        authorization.startsWith('Bearer ') && authorization.length > 7)) {
```
Comment (gannan08) [inline lib/http.js:282]: Where is the number 7 coming from? Can we declare a constant variable that tells us what this is? My assumption is that this is done to check that there is a token after `Bearer: ` in the authorization header?

Comment (mattcollier) [reply lib/http.js]: Attempted to address this in https://github.com/digitalbazaar/bedrock-vc-issuer/pull/29/commits/58f61818b1706a73032233c81546c34c2ff2b7ea

**Comment by dlongley** [comment]
@gannan08,

> In regards to issued credentials, I am unsure how we want to surface these credentials in the wallet -- if that's even something we want to do.

Nope, nothing we want to do with that right now. At the moment, the wallet doesn't know what's being stored in that EDV; it just has ultimate authority over it.

**Comment by dlongley** [comment]
Another option for parsing authorization headers: https://www.npmjs.com/package/auth-header

File: lib/http.js [hunk 2]
```diff
@@ -272,38 +273,31 @@ bedrock.events.on('bedrock-express.configure.routes', app => {
...
+      }
 
-      // get associated configuration
+      const token = authorization.substr(tokenStartIndex);
+      const profileAgentRecord = await profileAgents.getByToken(
+        {token, includeSecrets: true});
 
-      // TODO: a configuration will contain a schema that the incoming
-      // credential will be validated against
-      // const configuration = await vcIssuer.configurations.get({instance});
```
Comment (dlongley) [inline lib/http.js:292]: I'm concerned about losing these TODOs ... are we tracking elsewhere?

File: lib/issuer.js [hunk 3]
```diff
@@ -52,9 +64,107 @@ exports.issue = async ({instance, credential}) => {
...
+  const issuerKey = kms.createAsymmetricKey(
+    {invocationSigner, zcap: assertionMethodZcap});
+
+  const suite = new Ed25519Signature2018({
+    verificationMethod,
+    signer: issuerKey
+  });
+
+  // FIXME: How do is this supposed to work? Is it just a matter of
+  // removing the hash fragment?
```
Comment (dlongley) [inline lib/issuer.js:115]: The `issuer` is the `controller` of the verification method. We should have a better way of tracking this, however.

File: lib/issuer.js [hunk 4]
```diff
@@ -52,9 +64,107 @@ exports.issue = async ({instance, credential}) => {
...
+  const verifiableCredential = await vc.issue({
+    credential,
+    documentLoader,
+    suite
+  });
+
+  const edvClient = new EdvClient({
+    httpsAgent,
+    // FIXME: REQUIRED?
+    // id: edvId,
```
Comment (dlongley) [inline lib/issuer.js:128]: And `edvId` is not required, it just means the EdvClient will require capabilities to be passed when using it.

File: lib/issuer.js [hunk 5]
```diff
@@ -52,9 +64,107 @@ exports.issue = async ({instance, credential}) => {
...
+      invocationSigner,
+      kmsClient: new KmsClient({httpsAgent}),
+    })
+  });
+
+  // FIXME: need edvClient.ensureIndex, what are indexes on this collection?
+  // these indexes came from VerifiableCredentialStore.js in
+  // bedrock-web-vc-store
+  edvClient.ensureIndex({attribute: ['meta.issuer', 'content.type']});
+  edvClient.ensureIndex({attribute: 'content.id', unique: true});
```
Comment (dlongley) [inline lib/issuer.js:148]: I think these are ok for now. Let's role with it and adjust later. Ideally, we'll start expressing these in the profile's user doc that has some edv configuration info and we can just use that list, so it's all in one place.

File: lib/http.js (hunk 2 above)
Comment (dlongley) [reply lib/http.js:292]: I guess this one isn't that big of a deal and the others are handled or things have changed. Dismissing.

File: lib/http.js (hunk 1 above)
Comment (mattcollier) [reply lib/http.js]: Now using a lib and this code is eliminated.

File: lib/issuer.js (hunk 3 above)
Comment (mattcollier) [reply lib/issuer.js]: Added a did-resolver thing to use until did-io is where it needs to be.

File: lib/http.js [hunk 6]
```diff
@@ -273,22 +275,25 @@ bedrock.events.on('bedrock-express.configure.routes', app => {
...
-        authorization.length > tokenStartIndex)) {
-        throw new BedrockError('Permission denied.', 'NotAllowedError', {
-          httpStatusCode: 400,
-          public: true,
-        });
+      let token;
+      try {
+        // parse throws if the authorization header is missing
+        const auth = authorization.parse(req.get('authorization'));
+        if(!(auth.scheme === 'Bearer' && auth.token)) {
```
Comment (dlongley) [inline lib/http.js:282]: This is not a problem, just a note for you for the future: the `authorization` header is also used for http signatures, and at some point we'll likely enable this endpoint for use by frontends that use zcaps for authz. This is also probably a good reason for using a parser so we can check for multiple valid schemes.

**Pull request merged**